File: Backend/Application/DTO/Budget/Months/Recap/BudgetMonthRecapDto.cs
namespace Backend.Application.DTO.Budget.Months.Recap;

public sealed record BudgetMonthRecapDto(
    BudgetMonthRecapMetaDto Month,
    BudgetMonthRecapSnapshotTotalsDto SnapshotTotals,
    BudgetMonthRecapComparisonMetaDto Comparison,
    IReadOnlyList<BudgetMonthRecapExpenseCategoryDto> ExpenseCategories,
    BudgetMonthRecapSubscriptionInsightDto SubscriptionInsight,
    BudgetMonthRecapSavingsDetailDto SavingsDetail,
    BudgetMonthRecapDebtDetailDto DebtDetail);

public sealed record BudgetMonthRecapMetaDto(
    string YearMonth,
    string Status,
    DateTime OpenedAtUtc,
    DateTime? ClosedAtUtc,
    string CarryOverMode,
    decimal? CarryOverAmount);

public sealed record BudgetMonthRecapSnapshotTotalsDto(
    decimal TotalIncomeMonthly,
    decimal TotalExpensesMonthly,
    decimal TotalSavingsMonthly,
    decimal TotalDebtPaymentsMonthly,
    decimal FinalBalanceMonthly);

public sealed record BudgetMonthRecapComparisonMetaDto(
    string? PreviousComparableYearMonth,
    bool HasPreviousComparableMonth,
    BudgetMonthRecapComparisonSummaryDto? Summary);

public sealed record BudgetMonthRecapComparisonSummaryDto(
    BudgetMonthRecapMetricComparisonDto Income,
    BudgetMonthRecapMetricComparisonDto Expenses,
    BudgetMonthRecapMetricComparisonDto Savings,
    BudgetMonthRecapMetricComparisonDto DebtPayments,
    BudgetMonthRecapMetricComparisonDto FinalBalance);

public sealed record BudgetMonthRecapMetricComparisonDto(
    decimal PreviousValue,
    decimal DeltaAmount,
    decimal? DeltaPercent);

public sealed record BudgetMonthRecapExpenseCategoryDto(
    string CategoryId,
    string CategoryName,
    decimal CurrentAmount,
    decimal? PreviousAmount,
    decimal? DeltaAmount,
    decimal? DeltaPercent);

public sealed record BudgetMonthRecapSubscriptionInsightDto(
    IReadOnlyList<BudgetMonthRecapSubscriptionItemDto> Active,
    IReadOnlyList<BudgetMonthRecapSubscriptionItemDto> New,
    IReadOnlyList<BudgetMonthRecapSubscriptionItemDto> Removed,
    IReadOnlyList<BudgetMonthRecapSubscriptionItemDto> Paused,
    IReadOnlyList<BudgetMonthRecapSubscriptionItemDto> Cancelled,
    bool HasPreviousComparableMonth);

public sealed record BudgetMonthRecapSubscriptionItemDto(
    string IdentityKey,
    string Name,
    decimal AmountMonthly,
    string? SourceExpenseItemId);

public sealed record BudgetMonthRecapSavingsDetailDto(
    decimal TotalSavingsMonthly,
    IReadOnlyList<BudgetMonthRecapSavingsGoalDto> ActiveGoals,
    bool HasPreviousComparableMonth);

public sealed record BudgetMonthRecapSavingsGoalDto(
    string Id,
    string? SourceSavingsGoalId,
    string? Name,
    decimal MonthlyContribution,
    decimal? TargetAmount,
    DateTime? TargetDate,
    decimal? AmountSaved,
    decimal? PreviousMonthlyContribution,
    decimal? DeltaMonthlyContribution);

public sealed record BudgetMonthRecapDebtDetailDto(
    decimal TotalDebtPaymentsMonthly,
    IReadOnlyList<BudgetMonthRecapDebtItemDto> ActiveDebts,
    bool HasPreviousComparableMonth);

public sealed record BudgetMonthRecapDebtItemDto(
    string Id,
    string? SourceDebtId,
    string Name,
    string Type,
    decimal Balance,
    decimal Apr,
    decimal MonthlyPayment,
    decimal? MinPayment,
    decimal? MonthlyFee,
    int? TermMonths,
    decimal? PreviousMonthlyPayment,
    decimal? DeltaMonthlyPayment);

File: Frontend/src/types/budget/BudgetMonthRecapDto.ts
import type { BudgetMonthStatus, CarryOverMode } from "./BudgetMonthsStatusDto";

export type BudgetMonthRecapMetaDto = {
  yearMonth: string;
  status: BudgetMonthStatus;
  openedAtUtc: string;
  closedAtUtc: string | null;
  carryOverMode: CarryOverMode;
  carryOverAmount: number | null;
};

export type BudgetMonthRecapSnapshotTotalsDto = {
  totalIncomeMonthly: number;
  totalExpensesMonthly: number;
  totalSavingsMonthly: number;
  totalDebtPaymentsMonthly: number;
  finalBalanceMonthly: number;
};

export type BudgetMonthRecapComparisonMetaDto = {
  previousComparableYearMonth: string | null;
  hasPreviousComparableMonth: boolean;
  summary: BudgetMonthRecapComparisonSummaryDto | null;
};

export type BudgetMonthRecapComparisonSummaryDto = {
  income: BudgetMonthRecapMetricComparisonDto;
  expenses: BudgetMonthRecapMetricComparisonDto;
  savings: BudgetMonthRecapMetricComparisonDto;
  debtPayments: BudgetMonthRecapMetricComparisonDto;
  finalBalance: BudgetMonthRecapMetricComparisonDto;
};

export type BudgetMonthRecapMetricComparisonDto = {
  previousValue: number;
  deltaAmount: number;
  deltaPercent: number | null;
};

export type BudgetMonthRecapExpenseCategoryDto = {
  categoryId: string;
  categoryName: string;
  currentAmount: number;
  previousAmount: number | null;
  deltaAmount: number | null;
  deltaPercent: number | null;
};

export type BudgetMonthRecapSubscriptionInsightDto = {
  active: BudgetMonthRecapSubscriptionItemDto[];
  new: BudgetMonthRecapSubscriptionItemDto[];
  removed: BudgetMonthRecapSubscriptionItemDto[];
  paused: BudgetMonthRecapSubscriptionItemDto[];
  cancelled: BudgetMonthRecapSubscriptionItemDto[];
  hasPreviousComparableMonth: boolean;
};

export type BudgetMonthRecapSubscriptionItemDto = {
  identityKey: string;
  name: string;
  amountMonthly: number;
  sourceExpenseItemId: string | null;
};

export type BudgetMonthRecapSavingsDetailDto = {
  totalSavingsMonthly: number;
  activeGoals: BudgetMonthRecapSavingsGoalDto[];
  completedGoals: BudgetMonthRecapCompletedSavingsGoalDto[];
  hasPreviousComparableMonth: boolean;
};

// Snapshot of a goal that finished during this closed month. Display uses
// projectedAmountSaved (= raw amountSaved + monthlyContribution at close)
// as the canonical "reached this month" value — raw amountSaved is the
// stored, unmutated progression and must not be rendered as the final amount.
export type BudgetMonthRecapCompletedSavingsGoalDto = {
  id: string;
  name: string | null;
  targetAmount: number | null;
  amountSaved: number | null;
  monthlyContribution: number;
  projectedAmountSaved: number;
  closedAt: string;
};

export type BudgetMonthRecapSavingsGoalDto = {
  id: string;
  sourceSavingsGoalId: string | null;
  name: string | null;
  monthlyContribution: number;
  targetAmount: number | null;
  targetDate: string | null;
  amountSaved: number | null;
  previousMonthlyContribution: number | null;
  deltaMonthlyContribution: number | null;
};

export type BudgetMonthRecapDebtDetailDto = {
  totalDebtPaymentsMonthly: number;
  activeDebts: BudgetMonthRecapDebtItemDto[];
  hasPreviousComparableMonth: boolean;
};

export type BudgetMonthRecapDebtItemDto = {
  id: string;
  sourceDebtId: string | null;
  name: string;
  type: string;
  balance: number;
  apr: number;
  monthlyPayment: number;
  minPayment: number | null;
  monthlyFee: number | null;
  termMonths: number | null;
  previousMonthlyPayment: number | null;
  deltaMonthlyPayment: number | null;
};

export type BudgetMonthRecapExpenseDriverDto = {
  categoryId: string;
  categoryName: string;
  currentAmount: number;
  previousAmount: number;
  deltaAmount: number;
  deltaPercent: number | null;
};

export type BudgetMonthRecapInsightDriversDto = {
  expenseIncreaseDrivers: BudgetMonthRecapExpenseDriverDto[];
  largestExpenseIncreaseDriver: BudgetMonthRecapExpenseDriverDto | null;
};

// Explicit outcome of the carry-over decision recorded at close-time. The
// frontend should display carry-over from this DTO — `month.carryOverMode`
// and `month.carryOverAmount` are not reliable for `full` mode (Amount is
// stored as null on the next-month row).
export type BudgetMonthRecapCarryOverOutcomeDto = {
  mode: CarryOverMode;
  amount: number;
  targetYearMonth: string | null;
  wasApplied: boolean;
};

export type BudgetMonthRecapDto = {
  month: BudgetMonthRecapMetaDto;
  snapshotTotals: BudgetMonthRecapSnapshotTotalsDto;
  comparison: BudgetMonthRecapComparisonMetaDto;
  expenseCategories: BudgetMonthRecapExpenseCategoryDto[];
  subscriptionInsight: BudgetMonthRecapSubscriptionInsightDto;
  savingsDetail: BudgetMonthRecapSavingsDetailDto;
  debtDetail: BudgetMonthRecapDebtDetailDto;
  insightDrivers: BudgetMonthRecapInsightDriversDto;
  carryOverOutcome: BudgetMonthRecapCarryOverOutcomeDto;
};
